File: 2025-04-04/Island_Perimeter.java
package amazon;

public class Island_Perimeter {
    public static int islandPerimeter(int[][] grid) {
        int perimeter = 0;
        int[][] dirs = {{0,1},{0,-1},{1,0},{-1,0}};

        for(int i=0;i< grid.length;i++) {
            for(int j=0;j<grid[0].length;j++) {
                if(grid[i][j] == 0)
                    continue;

                for(int[] dir : dirs) {
                    int x = i+dir[0];
                    int y = j+dir[1];

                    if(x<0 || y<0 || x>=grid.length || y>=grid[0].length || grid[x][y] == 0)
                        perimeter++;
                }
            }
        }

        return perimeter;
    }
    public static void main(String[] args) {
        int[][] grid = {{0,1,0,0}, {1,1,1,0}, {0,1,0,0}, {1,1,0,0}};
        System.out.println(islandPerimeter(grid));
    }
}

File: 2026-03-05/Island_Perimeter.py
from typing import List


class Solution:
    def islandPerimeter(self, grid: List[List[int]]) -> int:
        n, m = len(grid), len(grid[0])
        perimeter = 0
        for i in range(n):
            for j in range(m):
                if grid[i][j]:
                    perimeter += 4

                    if i > 0 and grid[i-1][j]:
                        perimeter -= 2

                    if j > 0 and grid[i][j-1]:
                        perimeter -= 2

        return perimeter


sol = Solution()
print(sol.islandPerimeter([[0,1,0,0],[1,1,1,0],[0,1,0,0],[1,1,0,0]]))